### PR TITLE
systemd: allow only a single daemon-reload at the same time

### PR DESCRIPTION
**Pull request by mvo5**
This is an experiment to see if the "mount protocol error" reported in https://github.com/systemd/systemd/issues/10872 can be worked around by serializing the mount unit adding/removal. Proposing to get full spread runs.

### CODE REVIEW

File: overlord/snapstate/backend/mountunit.go [hunk 1]
```diff
@@ -32,7 +33,12 @@ import (
 	"github.com/snapcore/snapd/systemd"
 )
 
+var GIL = &sync.Mutex{}
```
[the remark this answers is not in the document]
Comment (mvo5) [reply overlord/snapstate/backend/mountunit.go]: Haha - I don't think this will land as is but I'm happy to change it (once the full test ran).

**Comment by mvo5** [comment]
This was green in https://travis-ci.org/snapcore/snapd/jobs/461360585

**Comment by mvo5** [comment]
And green again

**Comment by mvo5** [comment]
I just ran this on google:ubuntu-18.04-64 again and it passed there too.

**Comment by zyga** [comment]
I'm very optimistic about this :)

**Comment by mvo5** [comment]
Run it yet again in google:ubuntu-18.04-64 - still no error. I guess we can start thinking about how we can do something like this the right way(tm) now so that we can integrate it proper.

**Comment by zyga** [comment]
I think this is good to go. We want to have this, either in the current form or in some changed form down the line. LGTM

File: overlord/snapstate/backend/mountunit.go [hunk 2]
```diff
@@ -32,7 +33,12 @@ import (
 	"github.com/snapcore/snapd/systemd"
 )
 
+var GML = &sync.Mutex{}
```
Comment (stolowski) [inline overlord/snapstate/backend/mountunit.go]: Could you please add a comment describing the problem it solves?

**Comment by mvo5** [comment]
Needs more work as Samuele pointed out that we may need to disallow all daemon-reloads - its not clear if the nature of the bug is that there can be no extra mount units added (but multiple daemon-reloads are ok) or if when a mount unit is added only a single daemon reload is ok.

**Comment by pedronis** [comment]
This was the discussion @mvo5 mentioned:

<pedronis> mvo: my main comment there, thinking a bit further, is that the description is misleading, we do daemon-reload in other places?
<pedronis> mvo: if to fix it we really need to serialize daemon-reload then that is not enough, no?
<mvo> pedronis: let me see, I think you are right
<pedronis> so it would need more work, and doesn't feel a .3 thing
<mvo> pedronis: hm, in the overlord we really only do the daemon reload there afaict - but yeah, lets push it out
<pedronis> let me double check
<pedronis> mvo: we do reloads in wrappers which is used indirectly by overload for example
<pedronis> sorry, overlord
<mvo> pedronis: aha, its the only explicit place. indeed
<pedronis> also interfaces systemd backend does it
<pedronis> let me put this chat in the PR

**Comment by mvo5** [comment]
Closing this one as the rabbit hole is deeper and we need to ensure we don't do any daemon reloads in parallel with writing mount units. I will push a new PR soon.

**Pull request closed without merging**